### ISSUE TITLE
MLM训练的机制

### ISSUE THREAD

**Issue by dengkunn** [CyberZHG/keras-bert#127]
Hi, 您好！如果我们做预训练的MLM任务，论文中是要求只针对输入中被Mask的字所对应的输出求Loss，做梯度下降来优化，这样的话，应该是需要用到　masked_inputs这个输入，用０来遮蔽所对应字的输出；代码里似乎没有实现这一功能，代码似乎是针对所有非padding的字的loss来做优化的?　如果代码里有实现，麻烦帮指点下；或者不按照论文实现有其它的考虑？　谢谢！

**Comment by dengkunn** [comment]
Hi, 我找到了；如下设置成return_masked=True即可；
masked_layer = Masked(name='MLM', return_masked=True)([mlm_pred_layer, inputs[-1]])

同时构建model的时候，这个时候因为masked_layer 是个list，我们应该取第一个元素; masked_layer [0]，这个keras封装的tensor就把mask带上的，然后计算loss和梯度下降的时候，会自动使用这个tensor里的_keras_mask属性，不需要我们定制化loss；是这样的吗？谢谢！
model = keras.models.Model(inputs=inputs, outputs=[masked_layer [0], nsp_pred_layer])

**Comment by CyberZHG** [comment]
`return_masked=True`只用于单元测试，mask使用`compute_mask`的结果。

**Comment by dengkunn** [comment]
![Screenshot from 2019-08-22 10-11-34](https://user-images.githubusercontent.com/34706567/63480654-47941800-c4c5-11e9-9d56-79d548cd31d5.png)
Hi, 我的意思是如果我们要做预训练，return_masked需要设置成true是吧，我对比了设置与不设置的结果，只有设置成了true，其产生的mask才与论文的训练任务是一致的，只对masked的字的输出做训练，是这样的吗？

**Comment by CyberZHG** [comment]
```python
import keras
import keras.backend as K
import numpy as np
from keras_bert.layers import Masked


def loss(y_true, y_pred):
    return K.sum(y_true, axis=-1)


inputs = [keras.layers.Input((5,)), keras.layers.Input((5,))]
embed = keras.layers.Embedding(input_dim=2, output_dim=3, mask_zero=True)(inputs[0])

# masked = Masked(return_masked=True)([embed, inputs[1]])[0]
masked = Masked()([embed, inputs[1]])

model = keras.models.Model(inputs, masked)
model.compile(
    optimizer='sgd',
    loss=loss,
)

token_input = np.array([
    [1, 1, 1, 0, 0],
    [1, 1, 1, 1, 0],
])
mask_input = np.array([
    [0, 1, 0, 0, 0],
    [1, 0, 0, 0, 0],
])
outputs = np.arange(30).reshape((2, 5, 3))
print(model.evaluate([token_input, mask_input], outputs))
```

**Comment by dengkunn** [comment]
![Screenshot from 2019-08-22 12-54-27](https://user-images.githubusercontent.com/34706567/63487085-78cc1280-c4dc-11e9-87d8-f4ea1c8e0368.png)
![Screenshot from 2019-08-22 12-57-45](https://user-images.githubusercontent.com/34706567/63487086-79fd3f80-c4dc-11e9-969f-5889357279ff.png)
我试了你的代码，感觉还是要用return_masked=True，这样loss函数的值才是期望的，才正确按照input_mask的遮蔽了，否则就是按照padding来遮蔽的；谢谢！

**Comment by CyberZHG** [comment]
The loss should be 30.0 if the mask works properly. Otherwise, the loss will be 40.2857.

**Comment by CyberZHG** [comment]
Strangely, you got neither of the two results. Are you using the latest version of Keras?

**Comment by CyberZHG** [comment]
The calculation of losses with masks is a bit different between `Keras` and `tf.keras`.
The loss should be 6.0 if the mask works properly with `tf.keras`.

**Comment by dengkunn** [comment]
i am using tf2.0. Do I need to downgrade to tf1.14 or just keep tf2.0 by using return_masked=True?
the loss should be: (12 + 48)/2 = 30, i think so. but i can't repeat it. I can load the pre-trained model to do the sentiment analysis, the results are very good. Now i just want to try to train the pre-train model.

![Screenshot from 2019-08-22 15-58-33](https://user-images.githubusercontent.com/34706567/63497111-bc7f4600-c4f5-11e9-8c52-5c3a4487e010.png)

**Comment by CyberZHG** [comment]
I can reproduce your result only with `1.13.2`. `2.0.0-beta1` works fine to me.

The loss should be 6.0 if you're using `tf.keras`.

**Comment by dengkunn** [comment]
if i don't set the TF_KERAS=1, it means the tool will use Keras instead of tf.keras. then i will got the error " tf don't have placeholder", i will show you the detailed error some hours later. So i have to set TF_KERAS=1 to move forward. Anyway, using tf.keras is also correct? The only difference is loss maybe more smaller, divided by sequence length, other than the corresponding masked character length, which may results in the gradient more smaller. so i can move forward by using tf.keras and return_masked=True? Thanks

**Comment by dengkunn** [comment]
if i don't set TF_KERAS=1, the exact error is , so i switch to TF_KERAS=1.

~/tensorflow/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py in placeholder(shape, ndim, dtype, sparse, name)
    515         x = tf.sparse_placeholder(dtype, shape=shape, name=name)
    516     else:
--> 517         x = tf.placeholder(dtype, shape=shape, name=name)
    518     x._keras_shape = shape
    519     x._uses_learning_phase = False

AttributeError: module 'tensorflow' has no attribute 'placeholder'

**Comment by dengkunn** [comment]
It seems that it's the eager mode that makes the masked loss inconsistent. I was using tf2.0 yesterday, and the eager mode is defaulted, so the result was 28.2. Of course, the inconsistency is only for tf.keras. And if the tf is 2.0 and then using Keras directly, some errors will pop up. 

![Screenshot from 2019-08-23 18-13-00](https://user-images.githubusercontent.com/34706567/63585453-ba8ab500-c5d1-11e9-8a17-66c716f12378.png)
![Screenshot from 2019-08-23 18-10-37](https://user-images.githubusercontent.com/34706567/63585358-8911e980-c5d1-11e9-861f-43996cecf408.png)

**Comment by CyberZHG** [comment]
It's fixed in `0.77.0`.
